working/1426-extension-types/feature-specification-implements-rep.md

```diff
@@ -0,0 +1,268 @@
+inline superinterfaces have a member named `m`, but one or more of the
+interfaces of `R1 .. Rk` has a member named `m`. A compile-time error
+occurs if there exist `j1` and `j2` in `1 .. k` and a member name `m` such
+that `m` does not have a combined member signature for `R1
+.. Rk`. Otherwise the member signature of `m` is that combined member
+signature.
+
+Invocations of members declared by _DV_ or declared by an inline
+superinterface of _DV_ and not declared by any of `Rj`, `j` in `1..k` are
+unaffected by the fact that _DV_ implements `R1 .. Rk`.
+
+*This could be an invocation of a member declared by _DV_, or by any of its
+non-inline superinterfaces, or both, but the rules are unchanged.*
+
+Let `m` be a member name which is not declared by _DV_. Assume that the
+interface of `Rj` has a member named `m`. A compile-time error occurs if an
+inline superinterface of _DV_ also declares a member named `m`.
+
+Let `m` be a member name which is not declared by _DV_ and not declared by
+an inline superinterface of _DV_. Assume that the interface of `Rj` has a
+member named `m` with signature `s` *(this is the combined member signature
+that may depend on other types in `R1 .. Rk`)*. An invocation of `m` on a
+receiver of type `V` (or `V<T1 .. Ts>` if _DV_ is generic) is then treated
+as the same invocation, but with signature `s`.
+
+*In other words: Conflicts among superinterfaces are treated the same,
+whether it is an inline or a non-inline superinterface. In both cases, _DV_
+can resolve the conflict by redeclaring the given member. No override check
+is applied, any signature with the given name will resolve the conflict. If
+there is no conflict then _DV_ will "forward to" the members of `R0`.*
+
+*An implementation may choose to implicitly induce forwarding members into
+_DV_ in order to enable invocation of members of `R0`. However, such
+forwarding members must preserve the semantics of a direct invocation. In
+particular, if an invocation omits some optional parameters then the
+invocation of a member of `R0` must use the default value for that
+parameter of the actually invoked instance method, not a statically known
+value.*
+
+```dart
+class C {
+  int m([int i = 0]) => i;
+}
+
+class D extends C {
+  int m([int i = 42, j = -1]) => i;
+}
+
+inline class V implements C {
+  final C it;
+  V(this.it);
+}
+
+void main() {
+  V v = V(D());
+  // v.m(3, 4); // Compile-time error: `m` signature is from `C`.
+  v.m(); // Returns 42, not 0.
+}
+```
+
+It is an implementation specific behavior whether a closurization
+*(also known as a tear-off)* of an inline class instance member which is
+obtained from the interface of `R0` is a tear-off of the member of the
+representation object, or it is a tear-off of an implicitly induced
+forwarding method.
+
+*This makes no difference for the behavior of an invocation of the
+tear-off, but it does change the results returned by `==`.*
```

Comment:
Good catch, that is true!

I would tend to say that even this difference is allowed to be implementation specific. I think it should be OK for an implementation to avoid the forwarding functions entirely if possible, and it should also be possible to use forwarding functions as tear-offs (such tear-offs could be constant expressions "after desugaring")

In the case where the forwarder needs to preserve a default value which is not statically known I think we can rely on backends to use sentinel values, or on `foo$1` calling `forwardee.foo$1`, `foo$2` calling `forwardee.foo$2`, such that the forwarded call _actually_ passes the same number of actual arguments as the call site. Leaf had the impression that this could be implemented with a reasonable amount of work.

(We might also want to be in a good position to support faithful forwarding in a more general sense.)

(For `implemements X`, each invocation/tear-off of a member from the interface of `X` is known to be such an invocation/tear-off at the call site. If we actually allow the implementation to make the choice and, e.g., perform the invocation/tear-off directly on the forwardee, then the choice to make this implementation specific behavior could actually be quite helpful.)

So the main issue is actually that we could observe the forwarder based on the function object identity, and based on the run-time type of the function object.

I think this is still acceptable, so I changed the text to explain that the run-time type could depend on this choice (tear off the representation object member directly, or tear off a forwarding function).